File: app/src/androidTest/java/com/xi/android/kantukuang/test/TestModule.java
package com.xi.android.kantukuang.test;

import com.google.api.client.http.HttpRequest;
import com.google.api.client.http.HttpRequestFactory;
import com.google.api.client.http.HttpRequestInitializer;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.http.LowLevelHttpRequest;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.JsonObjectParser;
import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.client.testing.http.MockLowLevelHttpRequest;
import com.google.api.client.testing.http.MockLowLevelHttpResponse;
import com.google.inject.AbstractModule;
import com.google.inject.Inject;
import com.google.inject.Provides;
import com.google.inject.Scopes;
import com.google.inject.Singleton;

import java.io.IOException;


public class TestModule extends AbstractModule {

    @Override
    protected void configure() {
        bind(JsonFactory.class).to(JacksonFactory.class).in(Scopes.SINGLETON);
        bind(HttpTransport.class).to(MockHttpTransport.class).in(Scopes.SINGLETON);
    }

    @Provides
    @Singleton
    private JsonObjectParser provideJsonObjectParser(JsonFactory jsonFactory) {
        return new JsonObjectParser(jsonFactory);
    }

    @Provides
    @Singleton
    private HttpRequestInitializer provideHttpRequestInitializer(
            final JsonObjectParser jsonObjectParser) {
        return new HttpRequestInitializer() {
            @Override
            public void initialize(HttpRequest request) throws IOException {
                request.setParser(jsonObjectParser);
            }
        };
    }

    @Provides
    @Singleton
    private HttpRequestFactory provideHttpRequestFactory(HttpTransport transport,
                                                         HttpRequestInitializer initializer) {
        return transport.createRequestFactory(initializer);
    }

    public static class MockHttpTransport extends com.google.api.client.testing.http.MockHttpTransport {
        private MockLowLevelHttpResponse mResponse;
        @Inject
        private MockLowLevelHttpRequest mRequest;

        @Override
        public LowLevelHttpRequest buildRequest(String method, String url) throws IOException {
            mRequest.setResponse(mResponse);

            return mRequest;
        }

        public void setResponse(MockLowLevelHttpResponse mResponse) {
            this.mResponse = mResponse;
        }
    }
}

File: app/src/androidTest/scala/com/shen/xi/android/tut/test/TestModule.scala
package com.shen.xi.android.tut.test

import java.io.IOException

import com.google.api.client.http.{HttpRequest, HttpRequestInitializer, HttpTransport}
import com.google.api.client.json.jackson2.JacksonFactory
import com.google.api.client.json.{JsonFactory, JsonObjectParser}
import com.google.api.client.testing.http.{MockLowLevelHttpRequest, MockLowLevelHttpResponse}
import com.google.inject.{AbstractModule, Inject, Provides, Scopes, Singleton}


object TestModule {

  class MockHttpTransport extends com.google.api.client.testing.http.MockHttpTransport {
    private var mResponse: MockLowLevelHttpResponse = null
    @Inject
    private var mRequest: MockLowLevelHttpRequest = null

    @throws[IOException]
    override def buildRequest(method: String, url: String) = {
      mRequest.setResponse(mResponse)

      mRequest
    }

    def setResponse(mResponse: MockLowLevelHttpResponse) = this.mResponse = mResponse

  }

}

class TestModule extends AbstractModule {

  import com.shen.xi.android.tut.test.TestModule.MockHttpTransport


  override def configure() = {
    bind(classOf[JsonFactory]).to(classOf[JacksonFactory]).in(Scopes.SINGLETON)
    bind(classOf[HttpTransport]).to(classOf[MockHttpTransport]).in(Scopes.SINGLETON)
  }

  @Provides
  @Singleton
  private def provideJsonObjectParser(jsonFactory: JsonFactory) = new JsonObjectParser(jsonFactory)

  @Provides
  @Singleton
  private def provideHttpRequestInitializer(jsonObjectParser: JsonObjectParser) =
    new HttpRequestInitializer() {

      @throws[IOException]
      override def initialize(request: HttpRequest) =
        request.setParser(jsonObjectParser)

    }

  @Provides
  @Singleton
  private def provideHttpRequestFactory(transport: HttpTransport, initializer: HttpRequestInitializer) =
    transport.createRequestFactory(initializer)

}
